weblate/formats/tests/test_formats.py

```diff
@@ -328,6 +339,84 @@ def test_new_unit(self) -> None:
+        # structural sharing for some formats (notably XML based formats that
+        # store a DOM tree internally, e.g. AndroidFormat).
+        #
+        # This structural sharing can lead to mutations leaking back to the
+        # template unit. In the context of this test, `removenotes()` leaks back
+        # to the template.
+        #
+        # Unfortunately, we must accept this structural sharing for performance
+        # reasons, see:
+        # https://github.com/WeblateOrg/weblate/pull/11937#discussion_r1662166224
+        # (a probably better alternative would be to fix and use
+        # `buildFromUnit` in ttkit).
+        #
+        # It turns out, that in practice, the modifications from the structural
+        # sharing do not actually impact Weblate's observable behavior.
+        # This is likely because it round trips through files.
+        #
+        # We mimic the file round-tripping behavior in this test.
+
+        template_file = os.path.join(self.tempdir, f"test.tmpl.{self.EXT}")
+        main_file = os.path.join(self.tempdir, f"test.{self.EXT}")
+
+        # Create the template under test
+        # Add a new string with a note to the template file and save it.
+        shutil.copy(self.FILE, template_file)
+        template_storage = self.FORMAT(template_file, is_template=True)
+        template_unit = template_storage.new_unit(self.NEW_UNIT_KEY, "Source string")
+        template_unit.unit.addnote(self.NOTE_FOR_TEST)
+        template_storage.save()
+
+        # Add a new language and translate the new string.
+        self.FORMAT.add_language(main_file, Language.objects.get(code="cs"), self.BASE)
+        target_storage = self.parse_file(main_file, template=template_file)
+        target_unit, add = target_storage.find_unit(
+            self.NEW_UNIT_KEY, self.NEW_UNIT_KEY
+        )
+        self.assertTrue(add)
+        target_storage.add_unit(target_unit)
+
+        # Eagerly check that the target unit does not have notes.
+        # This is the point where checking that the template unit still has the
+        # notes would potentially fail (depending on the exact format).
+        self.assert_no_notes(target_unit)
+
+        target_unit.set_target("Translated string (CS)")
+        target_storage.save()
+
+        # Reload the storage to check notes were correctly written.
+        target_storage = self.parse_file(main_file, template=template_file)
+        target_unit, add = target_storage.find_unit(
+            self.NEW_UNIT_KEY, self.NEW_UNIT_KEY
+        )
+        self.assertFalse(add)
+
+        # Check we get the aggregated notes through the unit wrapper:
+        # We always (additionally) display the template notes to the user
+        # (if they are different from the target unit notes).
+        self.assertEqual(target_unit.notes.strip(), self.NOTE_FOR_TEST)
+
+        # Check there are no notes on the underlying unit.
+        self.assert_no_notes(target_unit)
+
```

Comment:
Please include test for side effects of shallow copy here. I don't think we can guarantee that it will be never hit inside Weblate, so better to be safe. Following should work:

1. Save template store
2. Parse it again
3. Verify that the note is there